Top content bar

bozojovicic opened the pull request:
Fixes #2252.

Top content bar

# Merge Checklist:
- [x] Code selfreview has been performed ([Best Practices](https://github.com/axa-ch-webhub-cloud/pattern-library/blob/develop/CONTRIBUTION.md#best-practices)).
- [x] Tests are sufficient.
- [x] Documentation is up to date.
- [x] Changelog is up to date.
- [x] Typescript typings for properties are up to date.
- [x] Designers approved changes or no approval needed.
- [x] Dependencies to other components still work.

Commit:
Top content bar

Ben-Zahler commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -46,13 +55,31 @@ class AXATopContentBar extends InlineStyles {
+    setTimeout(() => {
+      contBarEl.classList.add(hideClass);
+      fireCustomEvent('axa-top-bar-close', null, window);
+    }, 300);
```
why do we need a 300ms delay?

bozojovicic replied:
@Ben-Zahler 
I tried to add some animation into closing process although that was not the requirement and also it doesn't work for desktop-overlay so I will just remove this timeout and do the simple close.

Ben-Zahler commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -9,18 +95,20 @@
     min-height: 50px;
     vertical-align: middle;
-
-    text-align: center;
```
do these changes impact the behaviour of the "existing" notification bar? That may be already in use , if our new designs collide with the existing implementation, this must be discussed.

bozojovicic replied:
@Ben-Zahler 
If none of the top-content-bar attributes are set the only difference between new and old top-content-bar is the close functionality. The close icon is displayed and of course once you close it it will remain closed during that session.

<img width="726" alt="Screenshot 2021-12-28 at 11 21 27" src="https://user-images.githubusercontent.com/37440641/147556601-53f6b455-732e-42cd-86a4-921cdfbd4f0d.png">

If we should not display the close button for existing instances of top-content-bar then we need to know when to display it. In that case it will be hidden by default and we need some new attribute like "closable=true"
Ok like that or any other idea?

Ben-Zahler replied:
that sounds all ver good, I think adding closeable=true is a good idea

bozojovicic replied:
The attribute closeable added

markus-walther replied:
I notice @bozojovicic that you haven't really answered @Ben-Zahler 's original question here. You are deleting CSS that presumably served some purpose. Can you demonstrate why it has no impact on the look'n feel of the original top-content-bar? Or is this change blessed by the designers somehow? Or are you silently fixing some previously undetected shortcomings? Please address these questions here, so we know ;-)

bozojovicic replied:
Top content bar used to be a super-simple component that displays a text and a button. Now it has the close icon, the left icon, both of them are optional and it also has different design for mobile devices. I had to change some of the existing code to merge the new functionality into the existing component. But the text is still center aligned, not with text-align:center, but with justify-content:center, but visually the text is centered as requested in the design. Also per the new design the component is now 64px high and also the font-size is 18px, Ben said that is agreed with the designer. All in all I had to change some of the old code to support the new design.

Commit:
Top content bar

Commit:
Unit test fixed

Commit:
Top content bar - closeable

Commit:
Top content bar - documentation

Commit:
Unit test fixed

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -1,6 +1,7 @@
 # Top content bar
 
 Used as top of the page warning or info box that can show text and have a call to action button.
+It can also be display at the bottom of the page on mobile devices.
```
Please detail the conditions under which it will switch to bottom display! Here, it would suffice to say "(see below for the _stickymobile_ attribute).".

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,22 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
 
+### icon
+
+Using the string attribute `icon` as icon name, an icon will be rendered. To see the full list of possible icons, see the [axa-icon](https://github.com/axa-ch-webhub-cloud/pattern-library/blob/develop/src/components/10-atoms/icon/README.md) README.
```
Better: string-valued attribute...

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,22 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+### overlaydesktop
+
+Set the attribute `overlaydesktop` to `true` to display the bar on top of DOM element that comes after this element. Usually that would be the Keyvisual CTA component.
```
better: Boolean attribute.

Grammar: on top of **the** DOM element....

Question: What restrictions come with this feature, why does it have 'desktop' in the name? This may need further explaining here.

bozojovicic replied:
I did my best to explain it in the updated description.

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,22 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+Set the attribute `overlaydesktop` to `true` to display the bar on top of DOM element that comes after this element. Usually that would be the Keyvisual CTA component.
+
+### closeable
```
Google search of 'closeable' vs 'closable' shows the latter is far more frequent => adjust spelling!

bozojovicic replied:
I searched for the same thing and I've seen that both are used but did not check which version is used more. Anyway now it is "closable".

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -34,6 +40,10 @@ class AXATopContentBar extends InlineStyles {
       href: { type: String },
       variant: { type: String },
+      icon: { type: String },
+      stickymobile: { type: String },
```
Type error: should be **Boolean**

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -34,6 +40,10 @@ class AXATopContentBar extends InlineStyles {
       variant: { type: String },
+      icon: { type: String },
+      stickymobile: { type: String },
+      overlaydesktop: { type: String },
```
ditto

bozojovicic replied:
I changed stickymobile, overlaydesktop and closable to Boolean

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -34,6 +40,10 @@ class AXATopContentBar extends InlineStyles {
+      icon: { type: String },
+      stickymobile: { type: String },
+      overlaydesktop: { type: String },
+      closeable: { type: String },
```
ditto, plus 'closable' property name change.

markus-walther commented on src/components/20-molecules/top-content-bar/index.react.d.ts:
```diff
@@ -4,7 +4,11 @@ export type Variant = 'warning';
   variant?: Variant;
-  ctatext: string;
+  icon?: string;
+  stickymobile?: string;
```
string -> boolean, I think.

markus-walther commented on src/components/20-molecules/top-content-bar/index.react.d.ts:
```diff
@@ -4,7 +4,11 @@ export type Variant = 'warning';
-  ctatext: string;
+  icon?: string;
+  stickymobile?: string;
+  overlaydesktop?: string;
```
ditto.

markus-walther commented on src/components/20-molecules/top-content-bar/index.react.d.ts:
```diff
@@ -4,7 +4,11 @@ export type Variant = 'warning';
+  icon?: string;
+  stickymobile?: string;
+  overlaydesktop?: string;
+  closeable?: string;
```
ditto, plus closable.

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -13,13 +14,18 @@ import {
 class AXATopContentBar extends InlineStyles {
   static get tagName() {
-    return 'axa-top-content-bar';
+    return elTagName;
```
please revert this change - it's best practice to have a readable string here for all our components!

bozojovicic replied:
reverted

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -13,13 +14,18 @@ import {
 import styles from './index.scss';
+import fireCustomEvent from '../../../utils/custom-event';
 
+const elTagName = 'axa-top-content-bar';
```
Please delete this line, cf. line 28.

bozojovicic replied:
deleted

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -46,13 +56,28 @@ class AXATopContentBar extends InlineStyles {
   }
 
+  onClose() {
+    const contBarEl = this.shadowRoot.querySelector(elRootSelector);
```
With lit-html/element, we have an _efficient_, reactive HTML (re)rendering library, so we should not resort to CSS-based hiding. It's also bad for mobile performance, where high DOM-node count is critical (in your present solution, no UI is shown when closed, yet all the DOM nodes remain underlyingly).

A better approach (after deleting lines 76-77):
-  define a property `closed: {type: Boolean, reflect: true}`
-  `this.closed = true` here
- in line 131: `..., closed } = this;`
- line 132: `if (closed) return html ``;`

markus-walther replied:
If, using the better approach, one needs to refer to the closed state in component CSS, `:host([closed])` can be used.

See HTML (click on the triangle!) <details> element and its `open` attribute for an analogous case.

bozojovicic replied:
I did it as described in the first comment and it works fine. The 2nd comment totally confuses me, I don't see why we need :host([closed]) when it works well also without it.

markus-walther replied:
The 2nd comment was proactive - if you don't need it _here_, that's fine.

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -46,13 +56,28 @@ class AXATopContentBar extends InlineStyles {
+  onClose() {
+    const contBarEl = this.shadowRoot.querySelector(elRootSelector);
+    contBarEl.classList.add(hideClass);
+    fireCustomEvent('axa-top-bar-close', null, window);
```
this needs a comment: `// let *others* know that this component is now closed`

And the name should start with the exact component name, so it would be findable via `git grep`: `axa-top-content-bar-close`.

Third, our best practice is to fire this at the host element itself, and let event bubbling take it from there: `fireCustomEvent('axa-top-content-bar-close', null, this);`

bozojovicic replied:
Event names are now 'axa-top-content-bar-open' and 'axa-top-content-bar-close'.
I added couple of comments in this method to explain why these events are used.
I changed **window** with **this** for the events.

bozojovicic replied:
Event name is now 'axa-top-content-bar-close'.
Changed 'window' with 'this'
Added a detailed comment to onClose() method.

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+    line-height: 24px;
+  }
+
+  &.hide {
```
delete when better approach (see above) is used.

bozojovicic replied:
**hide** deleted

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +128,76 @@ class AXATopContentBar extends InlineStyles {
+    const rootClasses = {
+      'm-top-content-bar__overlay-desktop': overlaydesktop === 'true',
+      'm-top-content-bar__bottom-mobile': stickymobile === 'true',
+      hide: elementHidden,
```
I suggest to delete this line, and use the better approach instead.

bozojovicic replied:
done

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +128,76 @@ class AXATopContentBar extends InlineStyles {
+    };
+
+    const rootClasses = {
+      'm-top-content-bar__overlay-desktop': overlaydesktop === 'true',
```
After making overlaydesktop a Boolean property, you should simplify this to `overlaydesktop`.

bozojovicic replied:
done

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +128,76 @@ class AXATopContentBar extends InlineStyles {
+
+    const rootClasses = {
+      'm-top-content-bar__overlay-desktop': overlaydesktop === 'true',
+      'm-top-content-bar__bottom-mobile': stickymobile === 'true',
```
ditto for stickymobile.

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +128,76 @@ class AXATopContentBar extends InlineStyles {
+    };
+
+    const closeClasses = {
+      'hide-close': closeable !== 'true',
```
ditto, `!closable`.

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +128,76 @@ class AXATopContentBar extends InlineStyles {
             </div>
-          </axa-container>
+            ${
+              closeable !== 'true'
```
Simplified and more idiomatic: `closable ? html ``: html ``<div class="m-top-content-bar-empty"></div>`.

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+      top: 60px;
+      max-height: 0px;
+      width: 100%;
+      width: -moz-available;
```
What's the purpose of lines 23-25?? I would suggest to either drop them, or add justifying comments.

bozojovicic replied:
After all the changes I made they are useless. Will remove them.

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+
+  &.m-top-content-bar__bottom-mobile:not(.m-top-content-bar__overlay-desktop) {
+    @include breakpoint($mediaquery-mobile-up) {
+      position: static;
```
This is the first use of `position: static` in the entire pattern-library. Why needed?

bozojovicic replied:
Absolutely useless now. Will remove it.

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -9,18 +99,20 @@
-
-    padding-top: 5px;
-    padding-bottom: 5px;
+    padding: 16px 16px;
```
Isn't that equivalent to the shorter: `padding: 16px` ?!

bozojovicic replied:
yes :) fixed

markus-walther commented on src/components/20-molecules/top-content-bar/story.js:
```diff
@@ -7,7 +7,32 @@ import readme from './README.md';
+  'warning / error': 'warning',
+};
+
+const stickyMobileOptions = {
```
A bit overkill perhaps to name a single-use options bundle, but OK...

bozojovicic replied:
Could not decide if I should name it warning or error :) Will rename it to simple warning

markus-walther commented on src/components/20-molecules/top-content-bar/story.js:
```diff
@@ -29,16 +54,38 @@ export const TopContentBar = () => {
 
+  setTimeout(() => {
+    if (!window.sessionStorage.getItem('top-content-bar-closed')) {
+      window.dispatchEvent(new window.Event('axa-top-bar-open'));
```
The mantra for components is: Props down, events up. So, I suggest to invert the sessionStorage test, and if the top-content-bar-closed item exists in it, say this **after render(template,wrapper)** (assuming the 'better approach'):
`wrapper.querySelector('axa-top-content-bar').closed = true;`.

bozojovicic replied:
I explained in one of my previous comments why we need to use events to open or close the bar from axa-ch-core module

bozojovicic replied:
Done

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -46,13 +56,28 @@ class AXATopContentBar extends InlineStyles {
+
   firstUpdated() {
     this.inlineStyles('childStyles');
     const links = Array.prototype.slice.call(this.querySelectorAll('axa-link'));
```
In 2022, we don't need to take care of IE11 anymore, so could do an en-passant legacy fix here:
`const links = [...this.querySelectorAll('axa-link')];`

bozojovicic replied:
done

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+$mediaquery-mobile-up: 411px;
+$mediaquery-tablet-up: 991px;
```
New breakpoints and/or bypassing the existing SASS definitions need extra justifications.

Here, your 991px breakpoint is just 1 px off the existing $mediaquery-md-up, so the latter should be used.

bozojovicic replied:
done

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+$mediaquery-mobile-up: 411px;
```
Has this new breakpoint been discussed with the designers? The existing system, blessed by designers at the time AFAIK, has as smallest breakpoint $mediaquery-xs-up: // (min-width: 576px). What motivates going smaller here?

bozojovicic replied:
@markus-walther 
In the design https://www.figma.com/file/6zurYk3bJpzUg0H2THSxGF/AXA-UI-Kit?node-id=8744%3A5711 we have breakpoint: mobile < 411px and tablet < 991px
I agree that I can use existing $breakpoint-md:992px;
but for mobile I had to set the new breakpoint.
Should I use existing $breakpoint-xs:576px; instead of 411px ?

markus-walther replied:
I can't say what motivated the designers to specify an even smaller mobile breakpoint - can you ask if that was deliberate or an oversight?

If oversight, use $breakpoint-xs, otherwise your new breakpoint (which you'd keep local to this component).

Please record the designer's decision in the PR description, so we know.

Ben-Zahler commented:
@amir-ag @markus-walther can you review this? thanks!

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -56,9 +156,52 @@
+  background-color: $color-malachite;
+}
+.m-top-content-bar__container--attention {
+  background-color: #BC9D45;
```
Best practice is to use only symbolic colours. So, if you can find a nearby existing colour in 00-colors.scss, use it. If not, and the new colour is designer-approved, it would be best to add it to the /// Background and Illustration section of said 00-colors.scss (consider a name and enter it to https://colornames.org/color/bc9d45, perhaps).

bozojovicic replied:
I found here, and only here, https://hexcolorpedia.com/color/bc9d45/ that this color is called "underbrush" so I will add  $color-underbrush: #bc9d45 to colors.scss

markus-walther replied:
Great!

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,22 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+### stickymobile
+
+Set the attribute `stickymobile` to `true` to display the bar at the bottom of the screen on mobile devices with width < 411px.
```
better: Boolean attribute

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+  }
+
+  .m-top-content-bar__content-text {
+    margin: 1em 0px;
```
Can simplify 0px to 0.

bozojovicic replied:
fixed

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -1,5 +1,95 @@
+    display: none;
+  }
+
+  &.m-top-content-bar__overlay-desktop,
```
The README talks 'on top of DOM element that comes after this element' - how is that guaranteed here, what's the mechanism? Looking at the CSS that implements this, I don't understand how it works.

bozojovicic replied:
According to https://www.figma.com/file/6zurYk3bJpzUg0H2THSxGF/AXA-UI-Kit?node-id=8744%3A5711 the bar will be placed after the header. The top content bar with default configuration would be just displayed there and push the other content down. But if "overlay on dekstop" is enabled the bar would be just displayed over the component that comes after the header. That is usually some hero component, or keyvisual.
Here are some examples. Imagine that this is our page without the bar displayed
<img width="755" alt="page" src="https://user-images.githubusercontent.com/37440641/148053410-cd253b4d-5ef0-49fa-9138-bb3445fa8996.png">
When bar is displayed with overlaydesktop=false it will be displayed between the header and subheader
<img width="758" alt="page with bar" src="https://user-images.githubusercontent.com/37440641/148053516-dc2c5b0d-1e9c-49ba-ad75-67abd9ebefa1.png">
When bar is displayed with overlaydesktop=true it will be displayed after the header but on top of the subheader so when we close the bar the page will look exactly as on the first image without pushing any content up or down.
<img width="756" alt="page with bar overlayed" src="https://user-images.githubusercontent.com/37440641/148053666-60b1b089-8883-435e-b2da-f66fb7a215e7.png">

I'm also not happy with the CSS code that implements this. When I push the new code I will come back to this comment.

markus-walther replied:
I understand it better now. But technically, nothing guarantees the overlay property, as seen in this screenshot where I added some CSS to the subheader:
![Screen Shot 2022-01-06 at 09 38 13](https://user-images.githubusercontent.com/46926507/148923954-a486dfd6-0d92-4516-b2b7-02124fbdae2c.png)
:

markus-walther replied:
Apart from the problem with the overlay implementation, a more fundamental question is why a PL component should be given the power to suppress another, neighbouring component - doesn't this break the fundamental contract of a "component", to not draw outside of the box it has been given?

markus-walther replied:
One could argue that the orchestration of such suppression of neighbouring components belongs one level higher, to page CSS - and that it has no place in component CSS, since arguably a generic top-content-bar component should not encode narrowly specific usage scenarios.

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,22 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+### closeable
+
+Set the attribute `closeable` to `true` to display the Close button. Once the bar is closed it will not be displayed any more during that browser session.
+
```
When following the 'better approach' (see below), describe an additional Boolean property/attribute `closed` here.

bozojovicic replied:
Done

markus-walther commented on src/components/20-molecules/top-content-bar/child.scss:
```diff
@@ -1,5 +1,17 @@
-axa-top-content-bar > axa-link {
-  padding: 5px;
+axa-top-content-bar {
+  axa-link {
```
Why no longer immediate-dominance '>' ?

bozojovicic replied:
No idea why I changed this. Rollbacked.

markus-walther replied:
Good!

markus-walther commented on src/components/20-molecules/top-content-bar/child.scss:
```diff
@@ -1,5 +1,17 @@
+    padding: 5px;
+  }
+
+  a {
```
What motivates this additional child CSS? Could it break backwards compatibility?

bozojovicic replied:
The text displayed in the bar is entered in page properties in rich text area where the user can also add the links. That link is not axa-link but the ordinary a[href] so I had to add some CSS here to make it look like the link in the design. This will not affect the axa-link-aem child components, if there are any.

markus-walther commented on src/components/20-molecules/top-content-bar/child.scss:
```diff
@@ -1,5 +1,17 @@
+  }
+
+  a {
+    color: white;
```
That's not in our 0-colors.scss, but $color-prim-white is.

bozojovicic replied:
fixed

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -46,13 +56,28 @@ class AXATopContentBar extends InlineStyles {
     super();
     applyDefaults(this);
 
+    window.addEventListener('axa-top-bar-open', () => {
```
Using the better approach discussed at line 76, this event listener seems superfluous. top-component-bar-open is the default **state** of the component, doesn't feel semantically correct to model it as an **event**. I suggest to delete it.

bozojovicic replied:
I have to explain couple of things. The component's visibility depends on the sessionStorage ie. once the user closed it, we need to set some flag in sessionStorage to keep it closed during that browser session. We can access sessionStorage only from the axa-ch-core module where we listen for the event 'axa-top-content-bar-close' to set the sessionStorage flag
https://github.com/axa-ch-webhub-cloud/axa-ch-core/pull/76
The bar is initially opened, by default. We don't have any requirement if the bar should be initially opened or closed (before we close/open it with new "closed" property) so I added one more boolean attribute **initiallyclosed** to make this configurable.

bozojovicic replied:
And yes, I deleted the 'axa-top-bar-open' event and I use the "closed" property instead to show/hide the bar

markus-walther commented on src/components/10-atoms/icon/index.js:
```diff
@@ -14,6 +14,11 @@ import {
+  Info_outlineSvg,
+  WarningSvg,
+  Cloud_queueSvg,
+  CloseSvg,
```
Question: Does this increase the bundle size wherever axa-icon is used? By how much?

markus-walther replied:
I asked because axa-icon also allows bring-your-own SVGs (see section 'Icon as a SVG string' of the README), and so the alternative would be to put these new icons into axa-top-content-bar (which is rarely used). Knowing the size increase, if any, would help make a rational decision here.

bozojovicic replied:
I measured the size of icon bundle with and without new icons
/src/components/10-atoms/icon/dist/index.js

- from develop branch 56391 byte
- from this branch with all new icons 57693 bytes

bozojovicic replied:
@markus-walther let me know if this is OK or I should use <svg> directly
I also don't understand how using <svg> will reduce the build size when at the end the whole code ends up in the same build file /etc/clientlibs/axa/hub/podaempatternslibrary.js
Using separated <svg> strings would actually increase the size of this file at the end (in my opinion).

markus-walther replied:
OK, so that's a 2.3 % increase or + 1.27 KB. That's acceptable to me - thanks for supplying the data, so a data-driven decision is possible!

On the podaempatternslibrary.js question: this bundle is just an artifact of our current inability to deliver truly component-by-component bundles, which is a planned improvement to further reduce page weight. Then, each per-component bundle matters separately. Plus, and this is what was probably missing in your understanding, the icon bundle can get independently bundled in pods, which are not forced to use the same version from podaempatternslibrary.js. So, one can easily have multiple copies of the icon bundle on 1 page!

markus-walther reviewed (changes requested):
Hello @bozojovicic ,

thanks a lot for your first contribution to the pattern library!

Found a bunch of things, not unexpected for first time - so, don't worry and don't take it personal ;-)

Feel free to ask me on Slack or make a meeting if things are too unclear anywhere, although I tried my best.

Let's continue to collaborate to get this merged soon!

Cheers, Markus

Commit:
Top content bar - PR comment fixes

Commit:
Top content bar - Prettier

Commit:
Top content bar events fix

Commit:
Top content bar

bozojovicic commented:
@markus-walther @Ben-Zahler I think I answered and fixed all comments by now. For couple of things I first answered that it cannot be fixed because of this and that but in all such situations at the end I realized that it actually can be done exactly as Markus explained so I did it. I hope I deleted all such "can't do it" comments.
If you find some comments where I did not answer anything you can count that I fixed it. Just didn't want to write DONE 30 times.
One thing that is out of my power are these tests that fail randomly. After every commit I checked and every time some different test is failing. When I run them locally nothing fails.

Commit:
Top content bar

markus-walther commented on src/components/00-materials/styles/00-colors.scss:
```diff
@@ -33,6 +33,7 @@ $color-tosca: #914146 !default;
 $color-azalea: #e196aa !default;
 $color-cotton-candy: #fad6de !default;
 $color-apache: #ddbe65 !default;
+$color-underbrush: #bc9d45 !default;
```
Good!

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -1,6 +1,7 @@
 # Top content bar
 
 Used as top of the page warning or info box that can show text and have a call to action button.
+It can also be display at the bottom of the screen on mobile devices (see below for the stickymobile attribute).
```
Good!

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,30 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+### stickymobile
+
+Set the boolean attribute `stickymobile` to `true` to display the bar at the bottom of the screen on mobile devices with width < 411px.
```
Spelling: boolean -> Boolean (after Georg Boole, i.e. an adjective derived from a _name_). Please change everywhere.

bozojovicic replied:
Fixed

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,30 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+### initiallyclosed
+
+Set the boolean attribute `initiallyclosed` to `true` to make the bar is initially closed (on page load). By default it is initially opened ie. displayed. If the bar is initially closed it will be displayed after the check that the bar is not closed by the user during that browser session. Also if the bar is initially opened it will be automatically closed after the check that the bar is closed b ...
```
Grammar: "to make the bar is initially closed (on page load)." -> "to force the notification bar to appear closed on page load."

markus-walther replied:
I honestly don't understand the 2 sentences that follow. Please rework or drop them if not truly necessary.

Since closed state seems to sync with session state, I also have questions regarding the UX (flicker? how can users understand a closed bar coming back to life without user action, isn't the only sensible transition open->closed?).

markus-walther replied:
(See below discussion in the code where this is used for why I think this can be eliminated alltogether).

bozojovicic replied:
I removed the initiallyclosed attribute. "closed" is used now for this

markus-walther replied:
Great!

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -47,15 +55,25 @@ class AXATopContentBar extends InlineStyles {
   firstUpdated() {
     this.inlineStyles('childStyles');
-    const links = Array.prototype.slice.call(this.querySelectorAll('axa-link'));
+    const links = [...this.querySelectorAll('axa-link')];
```
Good!

markus-walther commented on src/components/20-molecules/top-content-bar/index.js:
```diff
@@ -103,23 +121,88 @@ class AXATopContentBar extends InlineStyles {
+      closed,
+    } = this;
+
+    if (!closed) {
```
Now - from reading lines 133-137 I can understand initiallyclosed (which does not participate in any logic elsewhere) - and prove that it is _not_ needed.

Here is a truth table of `closed` and the consequences for `render`:

closed   render
-----------------
false       content
true        empty

Thus, initiallyclosed can be eliminated in favour of <axa-top-content-bar closed ...>. 

Is there any situation where this alternative expression of the initially-closed state cannot be used?

If not, please delete initiallyclosed in code and documentation.

bozojovicic replied:
initiallyclosed removed

markus-walther commented on src/components/20-molecules/top-content-bar/README.md:
```diff
@@ -21,6 +22,30 @@ If no attribute `href` is set, it will render an `axa-button` and the `onClick`
+
+### overlaydesktop
+
+Set the boolean attribute `overlaydesktop` to `true` to display the bar on top of the DOM element that comes after this element. Usually that would be the Keyvisual CTA component. This feature applies only for the screen width larger than 411px. For mobile devices the bar will either be displayed as usually with `position:static` or at the bottom of the screen if the attribute `stickymobile` is e ...
```
Grammar, wording: "This feature applies only **to** non-mobile screen widths." + usually -> usual

bozojovicic replied:
I removed the "overlaydesktop" totally from the code. From JS, CSS and readme. Only stickymobile remains. If I understood it well, you agreed with Ben that this should be implemented in axa-ch-core CSS.

markus-walther replied:
"you agreed with Ben that this should be implemented in axa-ch-core CSS." Yes, that was the recommendation.

markus-walther commented on src/components/20-molecules/top-content-bar/index.scss:
```diff
@@ -56,9 +139,52 @@
+  background-color: $color-malachite;
+}
+.m-top-content-bar__container--attention {
+  background-color: $color-underbrush;
```
Good!

markus-walther reviewed (changes requested):
Hey Bojo / Ben,

great progress!

Some issues remain, the most fundamental one being the question of whether overlaydesktop is a good idea.

Looking forward to the final version, hopefully soon!

--Markus

Commit:
Top content bar - overlaydesktop and initiallyclosed removed

bozojovicic commented:
@markus-walther I don't understand why I can't reply on some comments so I will write it here. I removed initiallyclosed and overlaydesktop "features" from code. @Ben-Zahler will ask the designer if mobile breakpoint 411px is mandatory or we can use the standard $breakpoint-xs:576px;
I think all other comments are fixed/answered.

Commit:
Top content bar - use common mobile breakpoint

bozojovicic commented:
@markus-walther @Ben-Zahler the designer said that we can use the common mobile breakpoint.
The one for 411px is deleted.

markus-walther commented:
> @markus-walther @Ben-Zahler the designer said that we can use the common mobile breakpoint. The one for 411px is deleted.

Great.

markus-walther commented:
> @markus-walther I don't understand why I can't reply on some comments so I will write it here. I removed initiallyclosed and overlaydesktop "features" from code.

That's great.

Ben-Zahler committed:
WEBHUB-1100: new version of top notification bar

bozojovicic committed:
Merge pull request #2253 from Ben-Zahler/feature/top-content-bar

WEBHUB-1100: new version of top notification bar

Commit:
Top content bar - prettier

Commit:
Top content bar - package lock
